pkg/collector/corechecks/systemd/systemd_test.go

```diff
@@ -661,6 +675,114 @@ unit_names:
+	stats.On("GetUnitTypeProperties", mock.Anything, "unit2.service", dbusTypeMap[typeService]).Return(map[string]interface{}{
+		"CPUUsageNSec":     uint64(1),
+		"CPUAccounting":    true,
+		"MemoryCurrent":    uint64(1),
+		"MemoryAccounting": false,
+		"TasksCurrent":     uint64(1),
+	}, nil)
+	stats.On("GetUnitTypeProperties", mock.Anything, "unit3.service", dbusTypeMap[typeService]).Return(map[string]interface{}{
+		"CPUUsageNSec":     uint64(1),
+		"CPUAccounting":    true,
+		"MemoryCurrent":    uint64(1),
+		"MemoryAccounting": false,
+		"TasksCurrent":     uint64(1),
+	}, nil)
+	stats.On("GetUnitTypeProperties", mock.Anything, "unit4.socket", dbusTypeMap[typeSocket]).Return(getCreatePropertieWithDefaults(map[string]interface{}{
+		"NAccepted":    uint64(1),
+		"NConnections": uint64(1),
+		"NRefused":     uint64(1),
+	}), nil)
+	stats.On("GetVersion", mock.Anything).Return(systemdVersion)
+
+	tests := map[string]struct {
+		unitNames      string
+		unitRegexes    string
+		monitoredUnits []string
+	}{
+		"monitor_units_1_and_2": {"", `'unit1', 'unit2'`, []string{"unit1.service", "unit2.service"}},
+		"monitor_all_services":  {"", `'.+\.service$'`, []string{"unit1.service", "unit2.service", "unit3.service"}},
+		"monitor_all_units":     {"", `'.+'`, []string{"unit1.service", "unit2.service", "unit3.service", "unit4.socket"}},
+		// unit_regexes selects all services and unit_names selects the socket unit.
+		"unit_names_complement_unit_regex": {"unit4.socket", `'.+\.service$'`, []string{"unit1.service", "unit2.service", "unit3.service", "unit4.socket"}},
+		// unit_regexes excludes service unit1 and unit2, but unit_names selects them.
+		"unit_names_override_unit_regex": {"'unit1.service', 'unit2.service'", `[^12]`, []string{"unit1.service", "unit2.service"}},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Given
+			rawInstanceConfig := []byte(fmt.Sprintf(`
+unit_names: [%s]
+unit_regexes: [%s]
+`, test.unitNames, test.unitRegexes))
+			check := SystemdCheck{stats: stats}
+			senderManager := mocksender.CreateDefaultDemultiplexer()
+			check.Configure(senderManager, integration.FakeConfigHash, rawInstanceConfig, nil, "test")
+
+			mockSender := mocksender.NewMockSenderWithSenderManager(check.ID(), senderManager)
+			mockSender.On("Gauge", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return()
+			mockSender.On("ServiceCheck", mock.Anything, mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return()
+			mockSender.On("Commit").Return()
+
+			// When
+			check.Run()
+
+			// Then
+			mockSender.AssertCalled(t, "ServiceCheck", canConnectServiceCheck, servicecheck.ServiceCheckOK, "", []string(nil), mock.Anything)
+			for _, unitName := range test.monitoredUnits {
+				tags := []string{fmt.Sprintf("unit:%s", unitName)}
+				metrics := unitsToMetrics[unitName]
+				for _, metric := range metrics {
+					mockSender.AssertCalled(t, "Gauge", metric, mock.Anything, "", tags)
+				}
+			}
```

Comment:
I think that we need to assert that we don't report metrics for other units that are not in `test.monitoredUnits` otherwise, this test would pass if we're monitoring everything as well.

Reply:
sg, I pushed a tweak to do this, lmk what you think!

Reply:
Nice!